### PR TITLE
update backend

### PR DESCRIPTION
**Pull request by asrarbutt**
update

### CODE REVIEW

File: backend/src/main/java/capstone/myfinancemanager/manager/controller/TransactionController.java [hunk 1]
```diff
@@ -26,9 +31,15 @@ public List<TransactionDto> getAllTransactions() {
     @ResponseStatus(code = HttpStatus.CREATED)
     @PostMapping
     public ResponseEntity<TransactionDto> addTransaction(@RequestBody TransactionCreationDto transactionCreation) {
+
+        Principal principal = SecurityContextHolder.getContext().getAuthentication();
```
Comment (jamarob) [inline backend/src/main/java/capstone/myfinancemanager/manager/controller/TransactionController.java:35]: Spring will inject the principal if you include it in the method signature like so `addTransaction(Principal principal, @RequestBody TransactionCreationDto transactionCreation)`

File: backend/src/main/java/capstone/myfinancemanager/manager/controller/TransactionController.java [hunk 2]
```diff
@@ -26,9 +31,15 @@ public List<TransactionDto> getAllTransactions() {
...
+        if (principal.getName() == null) {
```
Comment (jamarob) [inline backend/src/main/java/capstone/myfinancemanager/manager/controller/TransactionController.java:36]: When we have no logged in user the principal should be null? Please verify. A better way would be to require authentication in the security config for `POST /transaction`.
Custom exception with speaking name is a good idea!

File: backend/src/main/java/capstone/myfinancemanager/manager/controller/TransactionController.java [hunk 3]
```diff
@@ -37,6 +48,15 @@ public ResponseEntity<Void> deletePlant(@PathVariable String id) {
         return new ResponseEntity<>(deleteSuccess ? HttpStatus.NO_CONTENT : HttpStatus.NOT_FOUND);
     }
 
+    @PutMapping("/update/{id}")
```
Comment (jamarob) [inline backend/src/main/java/capstone/myfinancemanager/manager/controller/TransactionController.java:51]: Not a resource oriented url (not restful). Combination of path and http verb is sufficient, please use just `@PutMapping("{id}")`

File: backend/src/main/java/capstone/myfinancemanager/manager/exceptions/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -47,4 +47,13 @@ public ResponseEntity<Map<String, Object>> handlePasswordNotMatchException() {
 
         return new ResponseEntity<>(responseBody, HttpStatus.BAD_REQUEST);
     }
+
```
Comment (jamarob) [inline backend/src/main/java/capstone/myfinancemanager/manager/exceptions/GlobalExceptionHandler.java:50]: Why is this commented out? Either delete it or uncomment. No dead code in main, please.

File: backend/src/main/java/capstone/myfinancemanager/manager/exceptions/UserNotLoggedIn.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package capstone.myfinancemanager.manager.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value = HttpStatus.BAD_REQUEST)
```
Comment (jamarob) [inline backend/src/main/java/capstone/myfinancemanager/manager/exceptions/UserNotLoggedIn.java:6]: Nice! But we have a more specific status code 401 UNAUTHORIZED comes to mind ;)

File: backend/src/main/java/capstone/myfinancemanager/manager/service/TransactionService.java [hunk 6]
```diff
@@ -61,4 +65,19 @@ public boolean deleteTransaction(String id) {
         }
         return false;
     }
+
+
+    public Transaction updateTransaction(String id, TransactionCreationDto transactionUpdate) {
+
+        Transaction transactionToUpdate = transactionRepo.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.FORBIDDEN));
```
Comment (jamarob) [inline backend/src/main/java/capstone/myfinancemanager/manager/service/TransactionService.java:72]: Good use of orElseThrow nut 403 FORBIDDEN means the authorized user is missing priviledge to perform the requested operation. 404 NOT FOUND would be a better fit as we can not update something that is not present.

File: backend/src/test/java/capstone/myfinancemanager/manager/controller/TransactionControllerIntegrationTest.java [hunk 7]
```diff
@@ -128,4 +131,91 @@ void deleteTransactionNotExists() throws Exception {
...
+                .pictureId("url2")
+                .isIncome(false)
+                .build();
+
+        mockMvc.perform(
+                        MockMvcRequestBuilders.put("/transactions/update/" + "1")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(objectMapper.writeValueAsString(transactionCreationDto))
+                )
+                .andExpect(status().is(403));
```
Comment (jamarob) [inline backend/src/test/java/capstone/myfinancemanager/manager/controller/TransactionControllerIntegrationTest.java:217]: 404, see controller

having a test for missing privilege would have revealed commented out code in controller advice ;)

File: backend/src/test/java/capstone/myfinancemanager/manager/service/TransactionServiceTest.java [hunk 8]
```diff
@@ -64,6 +66,26 @@ class TransactionServiceTest {
...
+            .description("Tanken")
+            .amount(27.0)
+            .transactionDate(Long.parseLong("1661866382913"))
+            .category("TestCategory")
+            .userEmail("test@test.com")
+            .isIncome(true)
+            .pictureId("url")
+            .build();
+
+    TransactionCreationDto transactionCreationDtoUpdate = TransactionCreationDto.builder()
```
Comment (jamarob) [inline backend/src/test/java/capstone/myfinancemanager/manager/service/TransactionServiceTest.java:79]: `TransactionCreationDto` and then transaction**Creation**Dto**Update** makes me wonder. Maybe we can find a better name? Is TransactionDto used in responses and TransactionCreationDto in requests? Then maybe TransactionResponseDto and TransactionRequestDto? Or maybe create an extra class for updates? 
Names are hard.. :)

File: backend/src/test/java/capstone/myfinancemanager/manager/service/TransactionServiceTest.java [hunk 9]
```diff
@@ -116,5 +138,18 @@ void deleteTransactionNotExistsTest() {
...
+
+        //when
+        when(randomUUIDGenerator.getRandomId()).thenReturn(randomTestId);
+        when(timestampService.now()).thenReturn(Instant.parse("2022-08-23T09:22:41.255023Z"));
+        when(transactionRepo.save(any())).thenReturn(transaction1);
+        when(transactionRepo.findById(transaction1.getId())).thenReturn(Optional.ofNullable(transaction1));
+
+        Transaction acutalTansaction = transactionService.updateTransaction(transaction1.getId(), transactionCreationDto1);
+
+        Assertions.assertEquals(transaction1.getId(), acutalTansaction.getId());
```
Comment (jamarob) [inline backend/src/test/java/capstone/myfinancemanager/manager/service/TransactionServiceTest.java:152]: This only checks that for id equality but not that the update was applied. Maybe compare the objects or verify that the repos save method was called with the correct object?

File: frontend/src/components/AddTransaction.tsx [hunk 10]
```diff
@@ -33,21 +33,46 @@ export default function AddTransaction(props: AddTransactionProps) {
     const [pictureId, setPictureId] = useState<string>("");
     const [description, setDescription] = useState<string>("");
     const [amount, setAmount] = useState<number>(0);
+    const [addNewtransaction, setAddNewtransaction] = useState<TransactionCreationDto>();
```
Comment (jamarob) [inline frontend/src/components/AddTransaction.tsx:36]: Maybe `transactionToAdd` or simply `newTransaction`. `addNewTransaction` sounds like function name

File: frontend/src/components/Header.tsx [hunk 11]
```diff
@@ -3,14 +3,13 @@ import {useContext} from "react";
...
 export default function Header() {
-    const {userLoggedIn, logout} = useContext(AuthContext);
-
+    const {loggedInUser, logout} = useContext(AuthContext);
 
     return (
 
         <header>
-            <MyNavBar loggedInUser={userLoggedIn} logout={logout}/>
+            <MyNavBar loggedInUser={loggedInUser} logout={logout}/>
```
Comment (jamarob) [inline frontend/src/components/Header.tsx:11]: If only the navbar uses these why not use the context inside the navbar? Saves one layer of indirection

File: frontend/src/components/MyCard.css [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    flex-direction: row;
+    justify-content: center;
+    align-content: center;
+    width: 80%;
+    padding-left: 1rem;
+
+    border-radius: 0 50px 50px 0;
+    box-shadow: 5px 5px 5px 5px rgba(0, 0, 0, 0.1);
+    margin-bottom: .5em;
+    margin-left: 1rem;
```
Comment (jamarob) [inline frontend/src/components/MyCard.css:18]: mix of rem and em, intentional?

File: frontend/src/components/MyCard.tsx [hunk 13]
```diff
@@ -29,35 +30,31 @@ type MyCardProps = {
...
-            <CardContent sx={{display: 'flex', flexDirection: 'row', justifyContent: 'space-between',}}>
-
-                <Typography sx={{fontSize: 14}} color="text.secondary">
-                    {props.allTransaction.description}
-                </Typography>
-
-                <Typography sx={{fontSize: 14}} color="text.secondary">
-                    {props.allTransaction.amount}
-                </Typography>
+        <div style={{display: "flex", justifyContent: "center", alignContent: "center", maxWidth: "50rem"}}>
```
Comment (jamarob) [inline frontend/src/components/MyCard.tsx:33]: No inline styles please, use a class

File: frontend/src/components/MyCardList.tsx [hunk 14]
```diff
@@ -1,39 +1,25 @@
...
+
             {
-                allTransactions.length !== 0 ? (
-                    allTransactions.map(t =>
-                        <MyCard allTransaction={t} deleteTransaction={deleteTransaction}/>
+                allTransaction.length !== 0 ? (
+                    allTransaction.map(t =>
+                        <MyCard key={t.id} allTransaction={t} deleteTransaction={deleteTransaction}/>
                     )
                 ) : (<h1>Keine Transaktion vorhanden</h1>)
```
Comment (jamarob) [inline frontend/src/components/MyCardList.tsx:19]: Only one h1 per page, maybe conditionally replace the heading text or use another element here. Maybe a p?

File: frontend/src/components/UpdateTransaction.tsx [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+import TransactionContext from "../context/transaction/TransactionContext";
+
+
+const StlyeEditIcon = styled(EditIcon)`
+
+  font-size: 2rem;
+ `;
+
+type UpdateTransactionProps = {
+    allTransaction: TransactionDto;
```
Comment (jamarob) [inline frontend/src/components/UpdateTransaction.tsx:33]: typo `allTransactions`

File: frontend/src/components/UpdateTransaction.tsx [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+                "category": category,
+                "transactionDate": convertDateToNumber(date),
+                "isIncome": isIncome,
+                "pictureId": pictureId,
+            }
+        )
+    }, [date, description, amount, isIncome, pictureId, category])
+
+    if (!props.allTransaction) {
+        return <>Nicht Gefunden</>
```
Comment (jamarob) [inline frontend/src/components/UpdateTransaction.tsx:64]: I think this has semantic meaning so a wrapping text would be justified

File: frontend/src/context/authentication/AuthContext.ts [hunk 17]
```diff
@@ -4,11 +4,12 @@ import {createContext} from "react";
 export interface IContext {
```
Comment (jamarob) [inline frontend/src/context/authentication/AuthContext.ts:4]: `IContext` is to generic to be a named export

File: frontend/src/context/transaction/TransactionProvider.tsx [hunk 18]
```diff
@@ -9,32 +11,18 @@ interface Param {
...
+    useEffect(() => {
+        getAllTransactions();
+    }, [])
+
+    const getAllTransactions = () => {
+        return axios.get("/transactions")
+            .then((response) => {
+                return response.data
+            })
+            .then(data => setAllTransaction(data))
```
Comment (jamarob) [inline frontend/src/context/transaction/TransactionProvider.tsx:25]:
```
.then((response) => response.data)
.then(setAllTransaction)
```
Short and sweet ;)

File: frontend/src/context/transaction/TransactionProvider.tsx [hunk 19]
```diff
@@ -9,32 +11,18 @@ interface Param {
 
 export default function TransactionProvider({children}: Param) {
 
-    const getAllTransactions = () => {
-        return axios.get("/transactions").then((response) => {
-            return response.data
-        })
-    }
+    const [allTransaction, setAllTransaction] = useState<TransactionDto[]>([]);
```
Comment (jamarob) [inline frontend/src/context/transaction/TransactionProvider.tsx:14]: typo plural

**Review by jamarob — COMMENTED** [review]
Nice! Left some comments.. mostly on readability and naming.

Some general advice:

pull requests: Try to give more context for the reviewer via a description of what was changed in the pr, at least a more descriptive name would be nice. Comit messages did not help that much ;)

backend: research meaning of status codes the names may be misleading of what the status code is intended for.

frontend: maybe decide in one css solution, mui, css files or styled components.

File: backend/src/test/java/capstone/myfinancemanager/manager/controller/TransactionControllerIntegrationTest.java [hunk 20]
```diff
@@ -128,4 +131,91 @@ void deleteTransactionNotExists() throws Exception {
...
+    }
+
+
 }
```
Comment (irenepapaspyratos) [inline backend/src/test/java/capstone/myfinancemanager/manager/controller/TransactionControllerIntegrationTest.java:221]: Leerzeile am Ende

File: frontend/src/GlobleTheme/customTheme.ts [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+        allVariants: {
+            wordWrap: 'break-word',
+        },
+        button: {
+            fontStyle: 'bold',
+            fontSize: 15,
+        }
+    },
+
+})
```
Comment (irenepapaspyratos) [inline frontend/src/GlobleTheme/customTheme.ts:66]: Leerzeile am Ende

File: frontend/src/components/MyCard.css [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+    display: flex;
+    flex-direction: column;
+    justify-content: center;
+    align-items: flex-end;
+    padding-right: 1rem;
+
+    width: 30%;
+    border-radius: 20px;
+}
```
Comment (irenepapaspyratos) [inline frontend/src/components/MyCard.css:74]: Leerzeile am Ende

**Pull request merged**